src/extractor/guidance/intersection_generator.cpp

```diff
@@ -577,9 +580,13 @@ IntersectionGenerator::GetActualNextIntersection(const NodeID starting_node,
     // to prevent endless loops
     const auto termination_node = node_based_graph.GetTarget(via_edge);
 
-    while (result.size() == 2 &&
-           node_based_graph.GetEdgeData(via_edge).IsCompatibleTo(
-               node_based_graph.GetEdgeData(result[1].turn.eid)))
+    // using a maximum lookahead, we make sure not to end up in some form of loop
+    const constexpr std::size_t max_lookahead = 10;
+    for (std::size_t lookahead = 0;
+         lookahead < max_lookahead && (result.size() == 2 &&
+                                       node_based_graph.GetEdgeData(via_edge).IsCompatibleTo(
+                                           node_based_graph.GetEdgeData(result[1].turn.eid)));
+         ++lookahead)
```

Comment:
Well, we could remember all nodes we see adding an additional data structure (e.g. hashmap) and check whether any node is repeated. Since we usually have a single traffic light to look at, this check is a simple way around broken modelings. At the same time, we don't have to allocate a dedicated `unordered_set` for every iteration.

If you feel that we should be on the save side for even situations with more than 10 traffic-lights/reasons not to compress on a single edge, we can go for an unordered_map, though.